subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/ExternalDependencyInstrumentingArtifactTransform.java

```diff
@@ -26,39 +26,33 @@
 import java.io.File;
 
 import static org.gradle.api.internal.initialization.transform.utils.InstrumentationTransformUtils.DEPENDENCIES_SUPER_TYPES_FILE_NAME;
-import static org.gradle.api.internal.initialization.transform.utils.InstrumentationTransformUtils.FILE_MISSING_HASH;
 import static org.gradle.api.internal.initialization.transform.utils.InstrumentationTransformUtils.METADATA_FILE_NAME;
-import static org.gradle.api.internal.initialization.transform.utils.InstrumentationTransformUtils.createNewFile;
-import static org.gradle.internal.classpath.TransformedClassPath.INSTRUMENTATION_CLASSPATH_MARKER_FILE_NAME;
-import static org.gradle.internal.classpath.TransformedClassPath.ORIGINAL_DIR_NAME;
-import static org.gradle.internal.classpath.TransformedClassPath.ORIGINAL_FILE_PLACEHOLDER_SUFFIX;
+import static org.gradle.api.internal.initialization.transform.utils.InstrumentationTransformUtils.isAnalysisMetadataDir;
 
 /**
- * Artifact transform that instruments external plugins with Gradle instrumentation.
+ * Artifact transform that instruments external artifacts with Gradle instrumentation.
  */
 @DisableCachingByDefault(because = "Instrumented jars are too big to cache")
 public abstract class ExternalDependencyInstrumentingArtifactTransform extends BaseInstrumentingArtifactTransform {
 
     @Override
     public void transform(TransformOutputs outputs) {
-        File inputDir = getInput().get().getAsFile();
-        if (inputDir.getName().equals(INSTRUMENTATION_CLASSPATH_MARKER_FILE_NAME)) {
-            return;
-        }
-
-        InstrumentationArtifactMetadata metadata = readArtifactMetadata(inputDir);
-        if (metadata.getArtifactHash().equals(FILE_MISSING_HASH)) {
-            execute(null, outputs, __ -> {});
-        } else {
-            String hash = metadata.getArtifactHash();
-            long contextId = getParameters().getContextId().get();
-            File originalArtifact = getParameters().getBuildService().get().getOriginalFile(contextId, hash);
-            execute(originalArtifact, outputs, __ -> writeOriginalFilePlaceholder(hash, outputs));
+        // We simulate fan-in behaviour:
+        // We expect that a transform before this one outputs two artifacts: 1. analysis metadata and 2. the original file.
+        // So if the input is analysis metadata we use it and create instrumented artifact, otherwise it's original artifact and we output that.
```

Comment:
❓ Do I understand correctly that we only create an instrumented artifact if analysis metadata is available, and if not, we pass the uninstrumented artifact through?

Reply:
So `analysis metadata` will always be available if jar exist.

So here what happens is:
```mermaid
flowchart TB
previousStep{{"Previous step\n(Artifact transform)"}}
instrumentationAndUpgrade{{"Instrumentation and upgrade\n(Artifact transform)"}} 
instrumentationAndUpgrade2{{"Instrumentation and upgrade\n(Artifact transform)"}} 
originalJar[["Original Jar"]]
originalJar2[["Original Jar"]]
metadata[["Metadata"]]
instrumentedJar[["Instrumented Jar"]]
finalClasspath[["Final classpath"]]

previousStep --> metadata --> instrumentationAndUpgrade --> instrumentedJar --> finalClasspath
previousStep --> originalJar --> instrumentationAndUpgrade2 --> originalJar2 --> finalClasspath
```

Basically previous step outputs two artifacts. And then we have this if:
if `Instrumentation and upgrade` detects that it's a metadata:
    it discovers original Jar via build service and outputs instrumented jar
else:
    it's just original jar and we just output it
 
Alternative approach would be, to not have that `else` case and pass original jar through. But basically we could find the original jar for the instrumented jar from the original classpath after instrumentation (this could also work, but had some quirks when I tried to allign two classpaths, so I went with this)